File: src/tools/database_user.ts
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { GraphQLClient } from 'graphql-request';
import { z } from 'zod';
import {
  GetCloudDatabaseClusterUsersDocument,
  CreateCloudDatabaseClusterUserDocument,
  ModifyCloudDatabaseClusterUserDocument,
  DeleteCloudDatabaseClusterUserDocument,
  type GetCloudDatabaseClusterUsersQuery,
  type CreateCloudDatabaseClusterUserMutation,
  type ModifyCloudDatabaseClusterUserMutation,
} from '../generated/graphql.js';

const DatabaseUserPermission = z.object({
  databaseName: z.string(),
  permission: z.enum(['READ_ONLY', 'READ_WRITE']),
  state: z.enum(['PRESENT', 'ABSENT']),
});

export function registerDatabaseUserTools(server: McpServer, client: GraphQLClient): void {
  server.registerTool(
    'nexaa_db_user_list',
    {
      description: 'List all users of a cloud database cluster',
      inputSchema: {
        clusterName: z.string(),
        clusterNamespace: z.string(),
      },
    },
    async ({ clusterName, clusterNamespace }) => {
      const data = await client.request<GetCloudDatabaseClusterUsersQuery>(
        GetCloudDatabaseClusterUsersDocument,
        {
          cloudDatabaseCluster: {
            name: clusterName,
            namespace: clusterNamespace,
          },
        },
      );
      return {
        content: [
          {
            type: 'text',
            text: JSON.stringify(data.cloudDatabaseCluster?.users, null, 2),
          },
        ],
      };
    },
  );

  server.registerTool(
    'nexaa_db_user_create',
    {
      description: 'Create a user in a cloud database cluster',
      inputSchema: {
        clusterName: z.string(),
        clusterNamespace: z.string(),
        username: z.string(),
        password: z.string().optional(),
        permissions: z.array(DatabaseUserPermission).default([]),
      },
    },
    async ({ clusterName, clusterNamespace, username, password, permissions }) => {
      const data = await client.request<CreateCloudDatabaseClusterUserMutation>(
        CreateCloudDatabaseClusterUserDocument,
        {
          userInput: {
            cluster: { name: clusterName, namespace: clusterNamespace },
            user: {
              name: username,
              password,
              state: 'PRESENT',
              permissions,
            },
          },
        },
      );
      return {
        content: [
          {
            type: 'text',
            text: JSON.stringify(data.cloudDatabaseClusterUserCreate, null, 2),
          },
        ],
      };
    },
  );

  server.registerTool(
    'nexaa_db_user_modify',
    {
      description: "Modify a database user's permissions",
      inputSchema: {
        clusterName: z.string(),
        clusterNamespace: z.string(),
        username: z.string(),
        permissions: z.array(DatabaseUserPermission),
      },
    },
    async ({ clusterName, clusterNamespace, username, permissions }) => {
      const data = await client.request<ModifyCloudDatabaseClusterUserMutation>(
        ModifyCloudDatabaseClusterUserDocument,
        {
          userInput: {
            cluster: { name: clusterName, namespace: clusterNamespace },
            user: {
              name: username,
              state: 'PRESENT',
              permissions,
            },
          },
        },
      );
      return {
        content: [
          {
            type: 'text',
            text: JSON.stringify(data.cloudDatabaseClusterUserModify, null, 2),
          },
        ],
      };
    },
  );

  server.registerTool(
    'nexaa_db_user_delete',
    {
      description: 'Delete a user from a cloud database cluster',
      inputSchema: {
        clusterName: z.string(),
        clusterNamespace: z.string(),
        username: z.string(),
      },
    },
    async ({ clusterName, clusterNamespace, username }) => {
      await client.request(DeleteCloudDatabaseClusterUserDocument, {
        userInput: {
          cluster: { name: clusterName, namespace: clusterNamespace },
          name: username,
        },
      });
      return {
        content: [
          {
            type: 'text',
            text: `User "${username}" deleted from cluster "${clusterName}".`,
          },
        ],
      };
    },
  );
}

File: tools/database_user.go
package tools

import (
	"context"
	"fmt"

	"github.com/mark3labs/mcp-go/mcp"
	"github.com/mark3labs/mcp-go/server"
	"github.com/nexaa-cloud/nexaa-cli/api"
)

func RegisterDatabaseUser(s *server.MCPServer) {
	s.AddTool(
		mcp.NewTool("nexaa_db_user_list",
			mcp.WithDescription("List all users of a cloud database cluster"),
			mcp.WithString("clusterName", mcp.Required()),
			mcp.WithString("clusterNamespace", mcp.Required()),
		),
		func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			result, err := withAPI(func(c *api.Client) ([]api.CloudDatabaseClusterUserResult, error) {
				return c.CloudDatabaseClusterUserList(api.CloudDatabaseClusterResourceInput{
					Name:      req.GetString("clusterName", ""),
					Namespace: req.GetString("clusterNamespace", ""),
				})
			})
			if err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			return jsonResult(result), nil
		},
	)

	s.AddTool(
		mcp.NewTool("nexaa_db_user_create",
			mcp.WithDescription("Create a user in a cloud database cluster"),
			mcp.WithString("clusterName", mcp.Required()),
			mcp.WithString("clusterNamespace", mcp.Required()),
			mcp.WithString("username", mcp.Required()),
			mcp.WithString("password"),
			mcp.WithArray("permissions", mcp.Items(dbUserPermItemSchema)),
		),
		func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			var args struct {
				ClusterName      string          `json:"clusterName"`
				ClusterNamespace string          `json:"clusterNamespace"`
				Username         string          `json:"username"`
				Password         *string         `json:"password"`
				Permissions      []dbUserPermArg `json:"permissions"`
			}
			if err := req.BindArguments(&args); err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}

			result, err := withAPI(func(c *api.Client) (api.CloudDatabaseClusterUserResult, error) {
				return c.CloudDatabaseClusterUserCreate(api.CloudDatabaseClusterUserCreateInput{
					Cluster: api.CloudDatabaseClusterResourceInput{
						Name:      args.ClusterName,
						Namespace: args.ClusterNamespace,
					},
					User: api.DatabaseUserInput{
						Name:        args.Username,
						Password:    args.Password,
						State:       api.StatePresent,
						Permissions: toDatabaseUserPermissions(args.Permissions),
					},
				})
			})
			if err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			return jsonResult(result), nil
		},
	)

	s.AddTool(
		mcp.NewTool("nexaa_db_user_modify",
			mcp.WithDescription("Modify a database user's permissions"),
			mcp.WithString("clusterName", mcp.Required()),
			mcp.WithString("clusterNamespace", mcp.Required()),
			mcp.WithString("username", mcp.Required()),
			mcp.WithArray("permissions", mcp.Required(), mcp.Items(dbUserPermItemSchema)),
		),
		func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			var args struct {
				ClusterName      string          `json:"clusterName"`
				ClusterNamespace string          `json:"clusterNamespace"`
				Username         string          `json:"username"`
				Permissions      []dbUserPermArg `json:"permissions"`
			}
			if err := req.BindArguments(&args); err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}

			cluster := api.CloudDatabaseClusterResourceInput{
				Name:      args.ClusterName,
				Namespace: args.ClusterNamespace,
			}
			user := api.DatabaseUserInput{
				Name:        args.Username,
				State:       api.StatePresent,
				Permissions: toDatabaseUserPermissions(args.Permissions),
			}

			result, err := withAPI(func(c *api.Client) (api.CloudDatabaseClusterUserResult, error) {
				return c.CloudDatabaseClusterUserModify(api.CloudDatabaseClusterUserModifyInput{
					Cluster: &cluster,
					User:    &user,
				})
			})
			if err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			return jsonResult(result), nil
		},
	)

	s.AddTool(
		mcp.NewTool("nexaa_db_user_delete",
			mcp.WithDescription("Delete a user from a cloud database cluster. WARNING: This is irreversible. Requires confirm: true."),
			mcp.WithString("clusterName", mcp.Required()),
			mcp.WithString("clusterNamespace", mcp.Required()),
			mcp.WithString("username", mcp.Required()),
			mcp.WithBoolean("confirm", mcp.Required(), mcp.Description("Must be true to confirm deletion")),
		),
		func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			clusterName := req.GetString("clusterName", "")
			clusterNS := req.GetString("clusterNamespace", "")
			username := req.GetString("username", "")
			if ok, cancelled := confirmDelete(req, "database user"); !ok {
				return cancelled, nil
			}
			_, err := withAPI(func(c *api.Client) (bool, error) {
				return c.CloudDatabaseClusterUserDelete(api.CloudDatabaseClusterUserResourceInput{
					Cluster: api.CloudDatabaseClusterResourceInput{Name: clusterName, Namespace: clusterNS},
					Name:    username,
				})
			})
			if err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			return mcp.NewToolResultText(fmt.Sprintf("User %q deleted from cluster %q.", username, clusterName)), nil
		},
	)
}
